### ISSUE TITLE
Questions on " Hessian is not positive definite"

### ISSUE THREAD

**Issue by sleepoctopus-max** [chjackson/msm#108]
Hi Professor Jackson,

Recently I came cross a problem when doing an analysis, where I apply multi-state model in a cohort of 8000 participants, each of them had 2 or 3 time points to observe disease states (5 stage and death, and only adjacent transitions are permitted), resulting in a long-format data of over 16000 lines.

[df_long.csv](https://github.com/user-attachments/files/18161189/df_long.csv)
Here's my code, with age, sex, edu and points set as covariates. 

Q <- rbind(
  c(-0.5 ,  0.25,    0,   0,      0,     0.25),
  c(0.166, -0.498, 0.166, 0,      0,     0.166),
  c(0    ,  0.166,-0.498, 0.166,  0,     0.166),
  c(0    ,  0    , 0.166, -0.498, 0.166, 0.166),
  c(0    ,  0    ,   0  , 0 ,     -0.5,  0.5),
  c(0   ,   0    ,   0 ,  0 ,     0,     0)
)
df_long1$age10<-df_long1$age/10
model.msm <- msm(state ~ time,
                 subject = id,
                 data = df_long1,
                 qmatrix = Q,
                 death = TRUE,
                 method = "BFGS",
                 center = F,
                 covariates = ~ age10+sex+edu+score,
                 control = list(trace=1, REPORT=1, fnscale = 10000, reltol = 1e-8,maxit=5000))

I frequently got the warning:
In msm(stage ~ time, subject = id, data = df_long, qmatrix = Q,  :
  Optimisation has probably not converged to the maximum likelihood - Hessian is not positive definite.

I have read your github website, and try to increase maxit or fnscale, or rescale age, but it does not work. Besides, the model took quite a long time to run (especially when I add the number of participants to . I wonder if there are better ways for optimisation? or if there is a problem with my msm model?

I have attached my data and code for your reference. I would appreciate any recommendations or instructions from you.
Thank you again for your time and for contributing such a valuable tool.

**Comment by chjackson** [comment]
Summarising the data with `statetable.msm(state, id, data=df_long1)` I see only a handful of observations in states 5 and 6, and you are trying to estimate 12 transition rates plus 12*4 covariate effects.  So I guess this is an identifiability issue that would not be solved by tweaking the optimizer. 

It seems to converge for me with no covariates in the model, so I would suggest starting with a small model, and building up gradually to diagnose where the weak information is.   Perhaps you could try models with covariates on some but not all transitions, and/or equality constraints on the parameters, if these would be plausible in your application.

**Comment by sleepoctopus-max** [comment]
Thanks a lot for your timely response!
As you mentioned,

> Perhaps you could try models with covariates on some but not all transitions, and/or equality constraints on the parameters, if these would be plausible in your application.

When I tried the former, I would set a  I wonder that if a specific transition is set to 0 in my Q matrix (e.g. transition 1-5 or 1-4),  do I still have to specify the covariates used for this transition in the covariate list?

**Comment by chjackson** [comment]
No, you only need to specify covariates for transitions that you are putting covariates on, and you don't need any "~1" formulae.   If the transition rate is always zero, then the idea of covariates on it would not make sense.

**Comment by sleepoctopus-max** [comment]
Thanks you very much, and it works! 
However, I'm still confused regarding your latter recommendation i.e. try models with equality constraints on the parameters.  
I have read Section 2.7, model 3, which said that "constraint= "  constrains the effect of covariate to be equal for the specified progression/recovery/death rates. I wondered that 
(1)how to determine the effect of this covariate was equal across these transitions, and why would this step help my model?
(2)if the optimisation problem has been solved by your former recommendation (try models with covariates on some but not all transitions), is the constraints step necessary? Or should I add equality constraints in the first place, since reducing the number of covariates in some transitions (with fewer observations) might be a little arbitrary?

Again, truly appreciate your time and patience in addressing my query!

**Comment by chjackson** [comment]
No statistical model represents the real world perfectly. The first model you have gotten to converge may not necessarily be the best model possible, or even a suitable model for your data and question.

Constraining effects gives a model with fewer parameters to estimate.   For example, constraining the death rates to be equal might help if there are only a small numbers of deaths from each state - then there may not be enough information to precisely distinguish different death rates from each state.

When choosing models you should consider what is plausible in your application, your data, and the estimates and confidence intervals from any models you have managed to fit.  You might also consider statistical model comparison measures such as likelihood ratio or AIC, and goodness of fit checks such as [prevalence](https://chjackson.github.io/msm/reference/prevalence.msm.html) or [survival](https://chjackson.github.io/msm/reference/plot.survfit.msm.html) plots, but consider what the purpose of your model is.  [This paper](https://journals.sagepub.com/doi/10.1177/0962280209105541) reviews a range of goodness of fit measures.

**Comment by sleepoctopus-max** [comment]
I will go through the model and the materials you provided carefully.
Very very grateful to your valuable recommendations and kind guidance!
